h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -329,9 +345,13 @@ private void gridSearch(Grid<MP> grid) {
     try {
       // Get iterator to traverse hyper space
       HyperSpaceWalker.HyperSpaceIterator<MP> it = _hyperSpaceWalker.iterator();
-      // Number of traversed model parameters
-      int counter = grid.getModelCount();
-      while (it.hasNext(model)) {
+
+      // This counter will be used as index to generate keys for corresponding `successful` models
+      int numberOfBuiltModels = grid.getModelCount();
+
+      int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models();
+
+      while (it.hasNext(model) && (it.max_models() == 0 || grid.getModelCount() < maxNumberOfModelsToBuild)) {
```

Comment:
(it.max_models() == 0 || grid.getModelCount() < maxNumberOfModelsToBuild)

that seems like an always true because is defined as

    maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models()

It would make sense to me if `maxNumberOfModelsToBuild = it.max_models() - numberOfBuiltModels`

Reply:
@michalkurka We need to take into an account situations when we proceed to build models from a new grid which happens to have the very same key as previous grid(s). Let me give you an example:
1) Cartesian Grid1 with hp space of size 6 was fully built. 
```
int numberOfBuiltModels = grid.getModelCount(); // will be 0 for Grid1

int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models(); // will be 0 + 6 for Grid1

while (it.hasNext(model) && (it.max_models() == 0 || grid.getModelCount() < 6)) { .... // where grid.getModelCount() belongs to [0,1,2,3,4,5] depending on iteration
```
Now we want(for whatever reason) to train second Grid2 with the same key and with hp of size 20. 
Let's consider two options:

a) Grid2 is again Cartesian:
```
int numberOfBuiltModels = grid.getModelCount(); // will be 6 as Grid1 was fully built

int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models(); // will be 6 + 20 for Grid2 = 26

while (it.hasNext(model) && (it.max_models() == 0 || grid.getModelCount() < 26)) { .... // where grid.getModelCount() belongs to [6,7,....25] depending on iteration
```
b) Grid2 is RandomDiscrete with max_models = 10 which is < hp space size:
```
int numberOfBuiltModels = grid.getModelCount(); // will be 6 as Grid1 was fully built

int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models(); // will be 6 + 10 for max_models from Grid2

while (it.hasNext(model) && (it.max_models() == 0 || grid.getModelCount() < 16)) { .... // where grid.getModelCount() belongs to [6,7,....15] depending on iteration number
```

All these scenarios I believe are supported properly. `maxNumberOfModelsToBuild` means max number of models grid with this key can have.

Reply:
Ok, let me ask a question: What will your code do if I run the train method twice?

```
>>> search_criteria = {'strategy': 'RandomDiscrete', 'max_models': 2, 'seed': 1}
>>> gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
...                           grid_id='gbm_grid1',
...                           hyper_params=gbm_params1,search_criteria=search_criteria)
>>> gbm_grid1.train(x=x, y=y,
...                 training_frame=train,
...                 validation_frame=valid,
...                 ntrees=100,
...                 seed=1)
gbm Grid Build progress: |████████████████████████████████████████████████████████████████████| 100%
>>> gbm_grid1.train(x=x, y=y,
...                 training_frame=train,
...                 validation_frame=valid,
...                 ntrees=100,
...                 seed=1)
gbm Grid Build progress: |████████████████████████████████████████████████████████████████████| 100%
>>> gbm_grid1
    col_sample_rate learn_rate max_depth sample_rate          model_ids             logloss
0               0.5        0.1         5         0.8  gbm_grid1_model_1  0.5552990538781676
1               0.5       0.01         3         1.0  gbm_grid1_model_2  0.6291962754031823
```

Is the result going to have 2 models or 4?

Also a reminder: one of the use cases is to stop grid search and then continue training. What is the behavior in that case?

Reply:
One technical note (small one): for large hyperspaces max_models returns Integer.MAX_VALUE - your math thus has an integer overflow problem.

Reply:
@michalkurka could you please provide `gbm_params1` in our code snippet so that I can answer your question correctly?

> One technical note (small one): for large hyperspaces max_models returns Integer.MAX_VALUE - your math thus has an integer overflow problem.

We can try to use Long type or BigDecimal :) , but honestly I believe we are quite far from abilities to compute hyperspaces of these sizes. For example if average time to compute single permutation from the grid takes 1 second - it will take 68 years to build a space of Integer.MAX_VALUE size and OOM will happen much much earlier according to our AutoML benchmark. I think we have many places in our project where we rely on a common sense from the user when he/she defines an input for the system. Anyway if user wants to train more then MAX_VALUE models he will run into an issue even without my math as .max_models() returns int right now.

Reply:
> We can try to use Long type or BigDecimal :) , but honestly I believe we are quite far from abilities to compute hyperspaces of these sizes.

What do you mean by that? Are you saying that H2O won't handle a hyperspace of the size of Integer.MAX_VALUE with say max_models = 2?

Reply:
@michalkurka What I was trying to say is that this is a quite unusual use case. And you were talking about case when max_models is not 2:

>for large hyperspaces max_models returns Integer.MAX_VALUE - your math thus has an integer overflow problem

We just can stretch this logic to any extent. Not even Long or BigDecimal type would be enough to use. What is your suggestion here then? we can take Math.min(x, Int.MAX_VALUE) for `maxNumberOfModelsToBuild`

Reply:
Not really. Almost all the APIs functions return longs, when they do not return longs just like the max_value method they convert to Integer.MAX_VALUE:

https://github.com/h2oai/h2o-3/blob/master/h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java#L400

This tells us:
- the code assumes everything can be `long` (and it is the responsibility of that code to not to exceed long because that is the API contract)
- where things collapse into `ints` it is handled

Resolving my comment could be eg. changing `int maxNumberOfModelsToBuild` to `long maxNumberOfModelsToBuild`. No harm done and the users who has a theoretically giant search space and wants to build a reasonable number of models won't end up with no models because of integer overflow.

Reply:
@michalkurka 
> @michalkurka could you please provide gbm_params1 in your code snippet so that I can answer your question correctly?

I will try to guess then. Probably your hyperspace was of a size 2. It is probably not 100% true as there is a chance that iterator from second grid returns  same permutations as a first grid. 
As far as I know we right now have a check whether models parameters were already built for a given permutation therefore we don't have repetitions in a built models within the grid.

Let's say our hyperspace is bigger than 2 ( for instance 20). And we run two grids with same key with `max_models = 2` for both grids. Then we will have from 2 to 4 built models as a results. But `maxNumberOfModelsToBuild` will be equal to 4 and we have an opportunity to skip 2 models due to this logic https://github.com/h2oai/h2o-3/blob/5478d2c6abc3c2b49c25176b762bcbe583724dee/h2o-core/src/main/java/hex/grid/GridSearch.java#L477

Reply:
@michalkurka 
> Not really. Almost all the APIs functions return longs, when they do not return longs just like the max_value method they convert to Integer.MAX_VALUE:

this is great. And I also doing this type of handling when it is really makes things safe. 

> the code assumes everything can be long (and it is the responsibility of that code to not to exceed long because that is the API contract)

who will stop us from providing a Map of hyperparameters with up to Integer.MAX_VALUE of keys and where every key points to an array of Integer.MAX_VALUE elements. This will overflow long in `computeMaxSizeOfHyperSpace`. 

> the users who has a theoretically giant search space and wants to build a reasonable number of models won't end up

For cartesian case we always set max_models to a size of hp space. It means that is user wants to build many of such grids..... he will sooner or later overflow and even long. 

Note: my code would not overflow if user wants to build a reasonable number of models, as it would mean that .max_models() was set to a reasonable number and this:
```
int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models();
``` 
would not overflow int even if hyperspace is giant

So my point was that with reasonable users even right now code is not going to overflow. We can do long - no problem. But still if you want you can break it with examples from above.

Reply:
Just tried your code and it produces 4 models instead of 2 (current release):

```
>>> # GBM hyperparameters
... gbm_params1 = {'learn_rate': [0.01, 0.1],
...                 'max_depth': [3, 5, 9],
...                 'sample_rate': [0.8, 1.0],
...                 'col_sample_rate': [0.2, 0.5, 1.0]}
>>> 
>>> search_criteria = {'strategy': 'RandomDiscrete', 'max_models': 2, 'seed': 1}
>>> 
>>> # Train and validate a cartesian grid of GBMs
... gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
...                           grid_id='gbm_grid1',
...                           hyper_params=gbm_params1, search_criteria=search_criteria)
>>> 
>>> gbm_grid1.train(x=x, y=y,
...                 training_frame=train,
...                 validation_frame=valid,
...                 ntrees=100,
...                 seed=1)
gbm Grid Build progress: |████████████████████████████████████████████████████████████████████| 100%
>>> gbm_grid1
    col_sample_rate learn_rate max_depth sample_rate          model_ids             logloss
0               0.5        0.1         5         0.8  gbm_grid1_model_1  0.5552990538781676
1               0.5       0.01         3         1.0  gbm_grid1_model_2  0.6291962754031823

>>> gbm_grid1.train(x=x, y=y,
...                 training_frame=train,
...                 validation_frame=valid,
...                 ntrees=100,
...                 seed=1)
gbm Grid Build progress: |████████████████████████████████████████████████████████████████████| 100%
>>> gbm_grid1
    col_sample_rate learn_rate max_depth sample_rate          model_ids             logloss
0               1.0        0.1         3         0.8  gbm_grid1_model_5  0.5537787805597868
1               0.5        0.1         5         0.8  gbm_grid1_model_1  0.5552990538781676
2               1.0        0.1         5         0.8  gbm_grid1_model_6  0.5567215084062702
3               0.5       0.01         3         1.0  gbm_grid1_model_2  0.6291962754031823
```

Reply:
>  We can do long - no problem.
OK

Reply:
@michalkurka and this result matches what I said:
>  Then we will have from 2 to 4 built models as a results.

Your second run did not choose the same permutations as your first run. That is why 2+2. But  you could get less if we skip some of them in the second run.

Reply:
reduce your space to 3 elements. And you should get 3 models

Reply:
@michalkurka my understanding that every time you call gbm_grid1.train() from the client you start new grid search but with the same key ( like in the unit tests here https://github.com/h2oai/h2o-3/blob/5478d2c6abc3c2b49c25176b762bcbe583724dee/h2o-algos/src/test/java/hex/grid/GridTest.java#L543 ). Everu new run creates new iterator we don't have a state. That is why we do skipping based on models stat are stored in DKV.

Reply:
But from the client's perspective I agree it looks like max_models should be applied to all .train() calls. But we need to store state then ( max_models from the first run)

Reply:
> Your second run did not choose the same permutations as your first run. That is why 2+2. But you could get less if we skip some of them in the second run.

The code is fully seeded and always produces 2 models on the current release.

```
import h2o
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.grid.grid_search import H2OGridSearch

h2o.init()

# Import a sample binary outcome dataset into H2O
data = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
test = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")

# Identify predictors and response
x = data.columns
y = "response"
x.remove(y)

# For binary classification, response should be a factor
data[y] = data[y].asfactor()
test[y] = test[y].asfactor()

# Split data into train & validation
ss = data.split_frame(seed = 1)
train = ss[0]
valid = ss[1]

# GBM hyperparameters
gbm_params1 = {'learn_rate': [0.01, 0.1],
                'max_depth': [3, 5, 9],
                'sample_rate': [0.8, 1.0],
                'col_sample_rate': [0.2, 0.5, 1.0]}

search_criteria = {'strategy': 'RandomDiscrete', 'max_models': 2, 'seed': 1}

# Train and validate a cartesian grid of GBMs
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                          grid_id='gbm_grid1',
                          hyper_params=gbm_params1, search_criteria=search_criteria)
gbm_grid1.train(x=x, y=y,
                training_frame=train,
                validation_frame=valid,
                ntrees=100,
                seed=1)
gbm_grid1.train(x=x, y=y,
                training_frame=train,
                validation_frame=valid,
                ntrees=100,
                seed=1)
gbm_grid1
```

Your code produces 4 models.

Reply:
@michalkurka let me investigate then. Probably I was relying on java tests and behaviour is different there.

Reply:
@michalkurka 
My solution was based on an assumption that we should append grids. Confusion started in this test https://github.com/h2oai/h2o-3/blob/9ab1cec03da7f674ace8fab9f56e077de0d8c98f/h2o-algos/src/test/java/hex/glrm/GLRMGridTest.java#L96 and particularly this assert here https://github.com/h2oai/h2o-3/blob/9ab1cec03da7f674ace8fab9f56e077de0d8c98f/h2o-algos/src/test/java/hex/glrm/GLRMGridTest.java#L154

This test actually means that we can run two grids with the same key and final number of models can be more then `max_models`. 
I even added parallel version of `testGridAppend` test here: https://github.com/h2oai/h2o-3/blob/5478d2c6abc3c2b49c25176b762bcbe583724dee/h2o-algos/src/test/java/hex/grid/GridTest.java#L441

We can reproduce this case in a client:
```
# GBM hyperparameters
      gbm_params1 = {'learn_rate': [0.01],
                     'max_depth': [3, 5],
                     'sample_rate': [0.8],
                     'col_sample_rate': [0.2]}
      gbm_params2 = {'learn_rate': [0.01],
                     'max_depth': [3, 5],
                     'sample_rate': [0.8],
                     'col_sample_rate': [0.1]} # difference is here

      search_criteria = {'strategy': 'Cartesian'}

      # Train and validate a cartesian grid of GBMs
      gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                                grid_id='gbm_grid1',
                                hyper_params=gbm_params1, search_criteria=search_criteria)
      gbm_grid1.train(x=x, y=y,
                      training_frame=train,
                      validation_frame=valid,
                      ntrees=100,
                      seed=1)

      gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                                grid_id='gbm_grid1',
                                hyper_params=gbm_params2, search_criteria=search_criteria)
      gbm_grid1.train(x=x, y=y,
                      training_frame=train,
                      validation_frame=valid,
                      ntrees=100,
                      seed=1)
      print(gbm_grid1)
```
will return 4 models even though `max_models = 2`.

So I think it is a bad design of iterators. Even though both iterators implements following method
```
/**
     * @return the maximum number of models to build
     */
    int max_models();
```
Cartesian implementation does not use it to stop anything. And also `HyperSpaceSearchCriteria` does not have `max_models` and `set_max_models`, as a consequence CartesianSearchCriteria does not have them as well....but Cartesian iterator does know about `max_models` for some reason.

But anyway, looks like it is how it is. I'm going to follow assumption that for RandomDiscrete case we should not get more models than `max_models` that was used in a constructor of the H2OGridSearch on a client(same for other clients). If `max_models` is being changed for different runs of the same grid key.... then we should not add them up but should use Math.max(max_models1, max_models_new,...)

Reply:
@michalkurka we should use .max_model() = 0 for Cartesian instead of setting it to the size of hp space I think. Which will be handled here
```
(it.max_models() == 0 || grid.getModelCount() < maxNumberOfModelsToBuild)
```

Reply:
@michalkurka We need to decide what will be an expected behaviour in case we run grid multiple times as there is some inconsistency in behaviours right now:
1) In your py snippet expectation is that we respect max_models regardless of number of times user runs grid.
2) In java unit tests, we can find test that assumes that we will append grids. [Here](https://github.com/h2oai/h2o-3/blob/46681848d670c197bc32dcd4cd4618b3b7cc8f9d/h2o-algos/src/test/java/hex/grid/GridTest.java#L412) for example we expect 2 models even though size of hp space is 1 ( max_model therefore also 1). 

I'm going to fix `testFailedParamsRetention` test with either Random search and corresponding max_models for two grids(for Cartesian search we should not expect second model to be added). Or, which is better I think, to make Cartesian's max_models return 0 ( which just disables early stopping)